### PR TITLE
Add an option in Istio CA to create certificates for webhook services.

### PR DESCRIPTION
**Pull request by yusuoh**

### CODE REVIEW

File: security/cmd/istio_ca/main.go [hunk 1]
```diff
@@ -229,6 +242,19 @@ func runCA() {
 		fatalf("failed to create secret controller: %v", err)
 	}
 
+	if opts.genWebhookSecrets {
+		err = util.GenerateWebhookSecrets(cs.CoreV1(), ca, []util.WebhookIdentity{
```
Comment (myidpt) [inline security/cmd/istio_ca/main.go]: I think it's better to run this as a go routine. This function could take some time that delays the cert provision for K8s service accounts.

File: security/cmd/istio_ca/main.go [hunk 2]
```diff
@@ -183,6 +188,14 @@ func init() {
 	flags.DurationVar(&opts.probeCheckInterval, "probe-check-interval", defaultProbeCheckInterval,
 		"Interval of checking the liveness of the CA.")
 
+	// Generate webhook secrets
+	flags.BoolVar(&opts.genWebhookSecrets, "generate-webhook-secrets", false,
```
Comment (myidpt) [inline security/cmd/istio_ca/main.go]: I don't think you need this boolean. Just use empty injector-webhook-secret to indicate it's disabled.

File: security/pkg/util/webhook.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+// WebhookIdentity stores the information about the webhook service.
+type WebhookIdentity struct {
+	Service   string
+	Secret    string
+	Namespace string
+}
+
+// GenerateWebhookSecrets generate key and cert for webhook services and stores them in K8S secrets
+func GenerateWebhookSecrets(client corev1client.CoreV1Interface, ca ca.CertificateAuthority, webhooks []WebhookIdentity) error {
+	caCert, _, _, _ := ca.GetCAKeyCertBundle().GetAllPem()
```
Comment (myidpt) [inline security/pkg/util/webhook.go]: nit: s/caCert/caCertPem

File: security/cmd/istio_ca/main.go [hunk 4]
```diff
@@ -183,6 +188,14 @@ func init() {
...
+		"Generate secrets for Istio webhook services.")
+	flags.StringVar(&opts.injectorWebhookService, "injector-webhook-service", "",
+		"Sidecar injector webhook service name.")
+	flags.StringVar(&opts.injectorWebhookSecret, "injector-webhook-secret", "",
+		"Sidecar injector webhook secret name.")
```
Comment (ayj) [inline security/cmd/istio_ca/main.go]: I'd like to see a e2e design for how this is going to work? e.g. how does this scale to multiple webhooks? How do we install/webhook after the CA is already started?

File: security/pkg/util/webhook.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+// Retry several times when create secret to mitigate transient network failures.
+func createSecretWithRetry(client corev1client.SecretInterface, secret *corev1.Secret) error {
+	var err error
+	for i := 0; i < secretCreationRetry; i++ {
```
Comment (ayj) [inline security/pkg/util/webhook.go]: This seems brittle. What's the recovery mode for dependent webhooks if this fails?

File: security/pkg/util/webhook_test.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	keyBytes, ok := data["key.pem"]
+	if !ok || !validateKey(t, keyBytes) {
+		t.Errorf("Invalid key.pem")
+	}
+}
+
+func validateKey(t *testing.T, key []byte) bool {
+	p, _ := pem.Decode(key)
+	if p == nil {
+		t.Error("Cannot decode the key.")
```
[the remark this answers is not in the document]
Comment (ldemailly) [reply security/pkg/util/webhook_test.go]: Error is fine when there is no format, why should it be Errorf
or you mean `t.Errorf("Cannot decode the key %v", key)` maybe?

File: security/pkg/util/webhook.go (hunk 5 above)
Comment (yusuoh) [reply security/pkg/util/webhook.go]: This is the same way that CA creates other secret. We can treat it as a fatal error and terminate the CA job.

File: security/cmd/istio_ca/main.go (hunk 4 above)
Comment (yusuoh) [reply security/cmd/istio_ca/main.go]: I think we only need to support two or three webhooks, so maybe we can just add them when needed. Or alternatively, we could have a list of strings as the flag. CA will store key, server cert and ca cert in the secret. The install tool can read from the secret for webhook registration, or the webhook service itself can do that if we implement the self-registration.

**Comment by yusuoh** [comment]
See #4216

**Pull request closed without merging**